mabwiser/treebandit.py

```diff
@@ -0,0 +1,171 @@
+# -*- coding: utf-8 -*-
+# SPDX-License-Identifier: Apache-2.0
+
+from copy import deepcopy
+from typing import Union, Dict, List, NoReturn, Optional, Callable
+
+import numpy as np
+from sklearn.tree import DecisionTreeClassifier
+
+from mabwiser.base_mab import BaseMAB
+from mabwiser.greedy import _EpsilonGreedy
+from mabwiser.linear import _Linear
+from mabwiser.popularity import _Popularity
+from mabwiser.rand import _Random
+from mabwiser.softmax import _Softmax
+from mabwiser.thompson import _ThompsonSampling
+from mabwiser.ucb import _UCB1
+from mabwiser.utils import reset, argmax, Arm, Num, _BaseRNG
+
+
+class _TreeBandit(BaseMAB):
+    def __init__(self, rng: _BaseRNG, arms: List[Arm], n_jobs: int, backend: Optional[str],
+                 lp: Union[_EpsilonGreedy, _Linear, _Popularity, _Random, _Softmax, _ThompsonSampling, _UCB1],
+                 tree_parameters: Dict):
+        super().__init__(rng, arms, n_jobs, backend)
+        self.lp = lp
+        self.tree_parameters = tree_parameters
+        self.tree_parameters["random_state"] = rng.seed
+
+        self.arm_to_tree = None
+        self.arm_to_rewards = None
+
+        # Initialize the arm expectations to nan
+        # When there are neighbors, expectations of the underlying learning policy is used
+        # When there are no neighbors, return nan expectations
+        reset(self.arm_to_expectation, np.nan)
+
+    def fit(self, decisions: np.ndarray, rewards: np.ndarray, contexts: np.ndarray = None) -> NoReturn:
+
+        # Reset the decision trees of each arm
+        self.arm_to_tree = {arm: DecisionTreeClassifier(**self.tree_parameters) for arm in self.arms}
+        self.arm_to_rewards = {arm: dict() for arm in self.arms}
+
+        if isinstance(self.lp, _ThompsonSampling) and self.lp.binarizer:
+            self.lp.is_contextual_binarized = False
+            rewards = self.lp._get_binary_rewards(decisions, rewards)
+            self.lp.is_contextual_binarized = True
+
+        # Calculate fit
+        self._parallel_fit(decisions, rewards, contexts)
+
+    def partial_fit(self, decisions: np.ndarray, rewards: np.ndarray, contexts: np.ndarray = None) -> NoReturn:
+
+        # Update rewards list at leaves
+        # Keep track of arms whose trees have not been fitted, and fit them separately
+        unfitted_arms = []
```

Comment:
You might not need to track of untrained arms. Technically, isn't it equivalent to rewards of that arm being empty. 

Alternatively, cannot we do this: 

- Loop thru arms, instead of decisions, as in the fit_arm
- Grab all the indices that matters for the current arm in one go
- Then, if (arm_to_reward[arm]) succeeds, as in the if-check in fit_arm(), we know that the arm is trained, we append the rewards as done here
- If the arm has no rewards recorded, (which can happen a) in the original fit() there was no data for this arm OR b) this is a new arm coming from _update_new_arm()), then first "fit()" the tree and then "save" its rewards.

Reply:
Yes, exactly! I thought of this too, but then remembered that this was our initial approach and we changed it to its current version because it'd be more efficient. Do we want to change it back?

Reply:
What does the unfitted arms get used for? It doesn't seem to be available outside of this function

Reply:
Yeah, I wasn't sure how did we take care the untrained bit back then. 

One thing to consider; if sklearn.apply(X) allows finding the leaf index of multiple contexts at once, that would be quite nice actually, and the above alternative (as in the old way) would work?

Thoughts?

Reply:
For the unfitted_arms, I can set it as a class attribute as @erstrong suggested and update it during _fit_arm and partial_fit and keep our current code. Or we can get rid of it and do it the way @skadio suggests. My only concern with the latter is that even though .apply() can work on multiple contexts, it'll end up giving us a list of leaves and we'll have to loop through them again. It might be cleaner to just keep an unfitted_arms list to keep track of when to update and when to fit from scratch.